### PR TITLE
Adds $lookup stage in aggregation

### PR DESCRIPTION
**Pull request by jheden**
Adds the Equality Match feature described at https://docs.mongodb.com/manual/reference/operator/aggregation/lookup/

### CODE REVIEW

File: mongomock/collection.py [hunk 1]
```diff
@@ -1649,6 +1649,18 @@ def _extend_collection(out_collection, field, expression):
                 if k == '$match':
                     out_collection = [doc for doc in out_collection
                                       if filter_applies(v, doc)]
+                elif k == '$lookup':
+                    foreign_name = stage['$lookup']['from']
```
Comment (pcorpet) [inline mongomock/collection.py]: If any of those fields is missing, it should raise a `pymongo.errors.OperationFailure: must specify 'from' field for a $lookup` instead of a simple `KeyError`.

File: mongomock/collection.py [hunk 2]
```diff
@@ -1649,6 +1649,18 @@ def _extend_collection(out_collection, field, expression):
...
+                    local_field = stage['$lookup']['localField']
```
Comment (pcorpet) [inline mongomock/collection.py]: Also note that you don't handle the `pipeline` parameter, so make sure to raise a `NotImplementedError` if pipeline is set.

File: mongomock/collection.py [hunk 3]
```diff
@@ -1649,6 +1649,18 @@ def _extend_collection(out_collection, field, expression):
...
+                    foreign_field = stage['$lookup']['foreignField']
+                    local_name = stage['$lookup']['as']
+                    foreign_collection = self.database.get_collection(foreign_name)
+                    for doc in out_collection:
+                        query = doc[local_field]
```
Comment (pcorpet) [inline mongomock/collection.py]: what if `local_field` is missing? Please catch that case.

File: mongomock/collection.py [hunk 4]
```diff
@@ -1649,6 +1649,18 @@ def _extend_collection(out_collection, field, expression):
...
+                        if type(doc[local_field]) is list:
+                            query = {'$in': query}
+                        matches = foreign_collection.find({foreign_field: query})
```
Comment (pcorpet) [inline mongomock/collection.py]: I'm not sure how `$lookup` should work, please add a link to the MongoDB page describing the feature to your PR. Also if some docs have some weird values in the local_field like `{'$exists': True}`.

File: mongomock/collection.py [hunk 5]
```diff
@@ -1649,6 +1649,18 @@ def _extend_collection(out_collection, field, expression):
                 if k == '$match':
                     out_collection = [doc for doc in out_collection
                                       if filter_applies(v, doc)]
+                elif k == '$lookup':
```
Comment (pcorpet) [inline mongomock/collection.py:1652]: For such a big new feature, we require that you add an end to end test to make sure that `pymongo` behaves like `mongomock`: add something in `test__mongomock.py` please.

**Review by pcorpet — CHANGES_REQUESTED** [review]
Yet another feature! Thanks Jonathan.

File: mongomock/collection.py (hunk 4 above)
Comment (jheden) [reply mongomock/collection.py]: I added the link with restriction in my original comment, thank you for the suggestion.
As for the weird values (specifically field names prefixed with `$`), see https://docs.mongodb.com/manual/reference/limits/#Restrictions-on-Field-Names

File: mongomock/collection.py (hunk 5 above)
Comment (jheden) [reply mongomock/collection.py:1652]: Since the testing strategy with `MultiCollection` only allows for testing against one collection at a time without major overhaul or hacking, I decided to run the `lookup` against the same collection.

File: mongomock/collection.py [hunk 6]
```diff
@@ -1650,13 +1650,25 @@ def _extend_collection(out_collection, field, expression):
                     out_collection = [doc for doc in out_collection
                                       if filter_applies(v, doc)]
                 elif k == '$lookup':
+                    for operator in ['let', 'pipeline']:
```
Comment (pcorpet) [inline mongomock/collection.py]: I personally prefer using tuples rather than list for such short inlined lists.

File: mongomock/collection.py [hunk 7]
```diff
@@ -1650,13 +1650,25 @@ def _extend_collection(out_collection, field, expression):
...
+                        if not isinstance(stage['$lookup'][operator], str):
+                            raise OperationFailure('arguments to $lookup must be strings')
                     foreign_name = stage['$lookup']['from']
                     local_field = stage['$lookup']['localField']
                     foreign_field = stage['$lookup']['foreignField']
                     local_name = stage['$lookup']['as']
                     foreign_collection = self.database.get_collection(foreign_name)
                     for doc in out_collection:
-                        query = doc[local_field]
+                        query = doc[local_field] if local_field in doc else None
```
Comment (pcorpet) [inline mongomock/collection.py]: I think that a simple `doc.get(local_field)` would do the equivalent.

File: mongomock/collection.py (hunk 5 above)
Comment (pcorpet) [reply mongomock/collection.py:1652]: Sounds good to me. Thanks for adding.

File: mongomock/collection.py [hunk 8]
```diff
@@ -1649,6 +1649,30 @@ def _extend_collection(out_collection, field, expression):
...
+                            raise NotImplementedError(
+                                "Although %s is a valid lookup operator for the "
+                                "aggregation pipeline, it is currently not "
+                                "implemented in Mongomock." % operator)
+                    for operator in ['from', 'localField', 'foreignField', 'as']:
+                        if operator not in stage['$lookup']:
+                            raise OperationFailure(
+                                "must specify '%s' field for a $lookup" % operator)
+                        if not isinstance(stage['$lookup'][operator], str):
+                            raise OperationFailure('arguments to $lookup must be strings')
```
Comment (pcorpet) [inline mongomock/collection.py]: Please add a blank space after this line to clearly show that you're done with the initial checks.

File: mongomock/collection.py [hunk 9]
```diff
@@ -1649,6 +1649,30 @@ def _extend_collection(out_collection, field, expression):
...
+                        if operator in stage['$lookup']:
+                            raise NotImplementedError(
+                                "Although %s is a valid lookup operator for the "
+                                "aggregation pipeline, it is currently not "
+                                "implemented in Mongomock." % operator)
+                    for operator in ['from', 'localField', 'foreignField', 'as']:
+                        if operator not in stage['$lookup']:
+                            raise OperationFailure(
+                                "must specify '%s' field for a $lookup" % operator)
+                        if not isinstance(stage['$lookup'][operator], str):
```
Comment (pcorpet) [inline mongomock/collection.py]: You should also raise an error `pymongo.errors.OperationFailure: FieldPath field names may not start with '$'.` if localField or foreignField starts with a `$`.

File: mongomock/collection.py [hunk 10]
```diff
@@ -1649,6 +1649,30 @@ def _extend_collection(out_collection, field, expression):
...
+                            raise OperationFailure('arguments to $lookup must be strings')
+                    foreign_name = stage['$lookup']['from']
+                    local_field = stage['$lookup']['localField']
+                    foreign_field = stage['$lookup']['foreignField']
+                    local_name = stage['$lookup']['as']
+                    foreign_collection = self.database.get_collection(foreign_name)
+                    for doc in out_collection:
+                        query = doc[local_field] if local_field in doc else None
```
Comment (pcorpet) [inline mongomock/collection.py]: Note that `local_field` could contain a `.`. I suggest that for now you just raise a `NotImplementedError` if you see this so that we can get your code merged. If someone else cares about using `.`, they will send another PR.

Also it's the same for `local_name`.

**Review by pcorpet — CHANGES_REQUESTED** [review]
Thanks for the changes. I found some other corner cases which might behave differently than MongoDB.

Also feel free to add yourself in the contributors in the README.md if you wish.

**Review by pcorpet — APPROVED** [review]
